src/Main.hs

```diff
@@ -261,6 +233,49 @@ routeCrumbs = (Some Route_Index :|) . \case
       Some r
     ]
 
+routeOgpMeta :: Config -> Text -> Route a -> OpenGraph
+routeOgpMeta cfg realmName route =
+    OpenGraph
+      { _openGraph_title = routeTitle realmName route
+      , _openGraph_url = URI.mkURI $ Config.baseUrl cfg <> Rib.routeUrl route
+      , _openGraph_author = Nothing
+      , _openGraph_description = description
+      , _openGraph_siteName = realmName <> " Archive"
+      , _openGraph_type = ogType
+      , _openGraph_image = firstMessage >>= _messageAvatarUrl
+      }
+  where
+    description :: Maybe Text
+    description = case route of
+      Route_Index ->
+        Nothing
+      Route_Stream stream StreamRoute_Index ->
+        Just $ _streamDescription stream
+      Route_Stream _ (StreamRoute_Topic _) ->
+        T.take 300 . stripHtml . _messageContent <$> firstMessage
+    ogType :: Maybe OGType
+    ogType = fromTopic $ \_ _ ->
```

Comment:
I just re-reviewed it, and have one comment. The `fromTopic` function could be obviated by simply casing the `route` here:

```
ogType = case route of 
  Route_Stream stream (StreamRoute_Topic topic) -> 
    Article { ... }
  _ -> Nothing
```
Once you have both `stream` and `topic` in that case block, you wouldn't need `fromTopic` at all, can directly use these values.

Reply:
:+1: dropped the fromTopic function and just used case matches in `ogType` and `firstMessage`